torch/csrc/jit/script/compiler.cpp

```diff
@@ -239,19 +236,40 @@ struct Environment {
     }
   }
 
+  SugaredValuePtr insertLoad(const std::string& name, TypePtr type) {
+    auto g = b->owningGraph();
+    auto load = g->insertNode(g->createLoad(name, type));
+    return std::make_shared<SimpleValue>(load->output());
+  }
+
+  void insertStore(const std::string& name, const SourceRange& loc, Value* v) {
+    auto g = b->owningGraph();
+    auto store = g->insertNode(g->createStore(name, v))->setSourceRange(loc);
+    type_table[name] = store->input()->type();
+  }
+
   SugaredValuePtr findInThisFrame(const std::string& name) {
     auto it = value_table.find(name);
     if (it != value_table.end()) {
       return it->second;
     }
+    auto it2 = type_table.find(name);
+    if (it2 != type_table.end()) {
+      return insertLoad(name, it2->second);
+    }
     return nullptr;
   }
 
   SugaredValuePtr findInParentFrame(const std::string& name) {
     return next ? next->findInAnyFrame(name) : nullptr;
   }
 
+  void setType(const std::string& name, TypePtr type) {
+    type_table[name] = std::move(type);
+  }
+
   SugaredValuePtr findInAnyFrame(const std::string& name) {
+    WithInsertPoint insert(b);
```

Comment:
This is confusing action at a distance. Why should the caller expect this to insert a Load instruction in the block associated with this frame? It should be inserted into the current block. I know that sometimes we are editing a block that has already been emitted. The WithInsertPoint calls should go at the site where this is happening not secretly here. This is a bug waiting to happen.